File: backend/app/schemas/document.py
"""Document request/response schemas."""
from datetime import datetime
from typing import Optional, List, Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.document import DocumentCategory, DocumentStatus


class DocumentCreate(BaseModel):
    """Schema for uploading a document."""

    category: DocumentCategory
    entity_type: Optional[str] = Field(
        None, description="Entity type (Deal, Quote, Vendor, etc.) or null for company docs"
    )
    entity_id: Optional[UUID] = Field(None, description="Entity ID or null for company docs")
    description: Optional[str] = Field(None, max_length=500)
    tags: Optional[List[str]] = Field(default_factory=list)


class DocumentResponse(BaseModel):
    """Full document response."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    company_id: UUID
    entity_type: Optional[str]
    entity_id: Optional[UUID]
    category: DocumentCategory
    storage_bucket: str
    storage_key: str
    original_filename: str
    file_size_bytes: int
    mime_type: str
    extracted_text: Optional[str] = Field(None, description="Raw extracted text (may be large)")
    parsed_data: Optional[dict[str, Any]] = Field(None, description="AI-extracted structured data")
    status: DocumentStatus
    ai_confidence_score: Optional[float] = Field(None, ge=0.0, le=1.0)
    error_message: Optional[str] = None
    description: Optional[str] = None
    tags: Optional[List[str]] = None
    created_at: datetime
    updated_at: datetime
    deleted_at: Optional[datetime] = None


class DocumentResponseWithoutText(BaseModel):
    """Document response without extracted_text (for lists)."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    company_id: UUID
    entity_type: Optional[str]
    entity_id: Optional[UUID]
    category: DocumentCategory
    original_filename: str
    file_size_bytes: int
    mime_type: str
    parsed_data: Optional[dict[str, Any]] = None
    status: DocumentStatus
    ai_confidence_score: Optional[float] = None
    error_message: Optional[str] = None
    description: Optional[str] = None
    tags: Optional[List[str]] = None
    created_at: datetime
    updated_at: datetime


class DocumentListResponse(BaseModel):
    """Paginated list of documents."""

    items: List[DocumentResponseWithoutText]
    total: int
    skip: int
    limit: int


class DocumentDownloadUrlResponse(BaseModel):
    """Response with presigned download URL."""

    url: str
    expires_in_minutes: int = 60
    filename: str

File: frontend/lib/types/document.ts
/**
 * Document management types for M4 AI Document Management
 */

export enum DocumentCategory {
  RFQ = "rfq",
  VENDOR_PROPOSAL = "vendor_proposal",
  CERTIFICATE = "certificate",
  MATERIAL_CERTIFICATE = "material_certificate",
  SPEC_SHEET = "spec_sheet",
  INVOICE = "invoice",
  PACKING_LIST = "packing_list",
  TEST_REPORT = "test_report",
  COMPANY_POLICY = "company_policy",
  TEMPLATE = "template",
  OTHER = "other",
}

export enum DocumentStatus {
  UPLOADING = "uploading",
  PROCESSING = "processing",
  COMPLETED = "completed",
  FAILED = "failed",
}

export interface Document {
  id: string;
  company_id: string;
  entity_type: string | null;
  entity_id: string | null;
  category: DocumentCategory;
  storage_bucket: string;
  storage_key: string;
  original_filename: string;
  file_size_bytes: number;
  mime_type: string;
  extracted_text?: string; // May be large, excluded from list responses
  parsed_data?: Record<string, unknown> | null;
  status: DocumentStatus;
  ai_confidence_score?: number | null;
  error_message?: string | null;
  description?: string | null;
  tags?: string[] | null;
  created_at: string;
  updated_at: string;
  deleted_at?: string | null;
}

export interface DocumentListItem {
  id: string;
  company_id: string;
  entity_type: string | null;
  entity_id: string | null;
  category: DocumentCategory;
  original_filename: string;
  file_size_bytes: number;
  mime_type: string;
  parsed_data?: Record<string, unknown> | null;
  status: DocumentStatus;
  ai_confidence_score?: number | null;
  error_message?: string | null;
  description?: string | null;
  tags?: string[] | null;
  created_at: string;
  updated_at: string;
}

export interface DocumentListResponse {
  items: DocumentListItem[];
  total: number;
  skip: number;
  limit: number;
}

export interface DocumentDownloadUrlResponse {
  url: string;
  expires_in_minutes: number;
  filename: string;
}

export interface DocumentUploadFormData {
  file: File;
  category: DocumentCategory;
  entity_type?: string;
  entity_id?: string;
  description?: string;
  tags?: string[];
}

/**
 * Helper to format file size for display
 */
export function formatFileSize(bytes: number): string {
  if (bytes === 0) return "0 Bytes";

  const k = 1024;
  const sizes = ["Bytes", "KB", "MB", "GB"];
  const i = Math.floor(Math.log(bytes) / Math.log(k));

  return Math.round((bytes / Math.pow(k, i)) * 100) / 100 + " " + sizes[i];
}

/**
 * Helper to get category display name
 */
export function getCategoryLabel(category: DocumentCategory): string {
  const labels: Record<DocumentCategory, string> = {
    [DocumentCategory.RFQ]: "Request for Quote",
    [DocumentCategory.VENDOR_PROPOSAL]: "Vendor Proposal",
    [DocumentCategory.CERTIFICATE]: "Certificate",
    [DocumentCategory.MATERIAL_CERTIFICATE]: "Material Certificate",
    [DocumentCategory.SPEC_SHEET]: "Specification Sheet",
    [DocumentCategory.INVOICE]: "Invoice",
    [DocumentCategory.PACKING_LIST]: "Packing List",
    [DocumentCategory.TEST_REPORT]: "Test Report",
    [DocumentCategory.COMPANY_POLICY]: "Company Policy",
    [DocumentCategory.TEMPLATE]: "Template",
    [DocumentCategory.OTHER]: "Other",
  };

  return labels[category] || category;
}

/**
 * Helper to get status display name with color
 */
export function getStatusLabel(
  status: DocumentStatus
): { label: string; color: string } {
  const statusMap: Record<
    DocumentStatus,
    { label: string; color: string }
  > = {
    [DocumentStatus.UPLOADING]: {
      label: "Uploading",
      color: "bg-blue-100 text-blue-800",
    },
    [DocumentStatus.PROCESSING]: {
      label: "Processing",
      color: "bg-yellow-100 text-yellow-800",
    },
    [DocumentStatus.COMPLETED]: {
      label: "Completed",
      color: "bg-green-100 text-green-800",
    },
    [DocumentStatus.FAILED]: { label: "Failed", color: "bg-red-100 text-red-800" },
  };

  return statusMap[status] || { label: status, color: "bg-gray-100 text-gray-800" };
}
